devlib/connection.py

```diff
@@ -61,17 +62,128 @@ class ConnectionBase(InitCheckpoint):
     """
     Base class for all connections.
     """
+
+    _MAX_BACKGROUND_CMD = 50
+    """
+    Check for up to this amount of available background commands.
+    """
+
     def __init__(self):
         self._current_bg_cmds = WeakSet()
         self._closed = False
         self._close_lock = threading.Lock()
         self.busybox = None
+        self._bg_spawn_lock = threading.RLock()
 
     def cancel_running_command(self):
         bg_cmds = set(self._current_bg_cmds)
         for bg_cmd in bg_cmds:
             bg_cmd.cancel()
 
+    @property
+    def _bg_async_sem(self):
+        # If we have not installed busybox yet, we won't be able to poll for
+        # the max number of background commands so we return a dummy semaphore
+        # to force falling back on the blocking path.
+        if self.busybox is None:
+            return asyncio.BoundedSemaphore(0)
+        else:
+            return self._get_bg_async_sem()
+
+    # Memoization ensures we will get the same semaphore back all the time.
+    @memoized
+    def _get_bg_async_sem(self):
+        # Ensure we have a running asyncio loop, otherwise the semaphore will
+        # be useless
+        asyncio.get_running_loop()
+        return asyncio.BoundedSemaphore(self._max_async_bg)
+
+    @property
+    @memoized
+    def _max_bg(self):
```

Comment:
We technically could but I think it's quite dangerous. If someone puts e.g. 15 when the real value is 10 "everything will work fine" until the client code runs a bunch of commands in parallel, at which point something bad will happen. It could be some sort of generic exception with 0 information to allow debugging, or possibly a misleading exception, or even why not just hanging. In any case, we cannot expect any indication of the actual problem and the location of the exception will be unhelpful.

What could be done is expose as a setting the max number of polled connection (possibly 0) so the user can reduce the time spent in the setup phase if they are ok with sequential code. This would be totally safe in all circumstances.